### PR TITLE
[Feature/extensions] Added unit test for createComponent workflow

### PR DESCRIPTION
**Pull request by owaiskazi19**
Signed-off-by: Owais Kazi <owaiskazi19@gmail.com>

### Description
Added Unit tests for registerHandler of ClusterState, ClusterSettings and LocalNode
 
### Issues Resolved
Part of https://github.com/opensearch-project/opensearch-sdk/issues/25
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by owaiskazi19** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/105/
> * **CommitID:** [3f6b8b8](https://github.com/opensearch-project/OpenSearch/commit/3f6b8b84a37a392e639454f20c0e56a661ebe186)

@peterzhuamazon bot commented before the gradle check is still running. Any idea here?

**Comment by peterzhuamazon** [comment]
> > ### Gradle Check (Jenkins) Run Completed with:
> > 
> > * **RESULT:** FAILURE ❌
> > * **URL:** https://build.ci.opensearch.org/job/gradle-check/105/
> > * **CommitID:** [3f6b8b8](https://github.com/opensearch-project/OpenSearch/commit/3f6b8b84a37a392e639454f20c0e56a661ebe186)
> 
> @peterzhuamazon bot commented before the gradle check is still running. Any idea here?
Hi Owais, you force pushed the branch so it takes both


@[owaiskazi19](https://github.com/owaiskazi19) owaiskazi19 [force-pushed](https://github.com/opensearch-project/OpenSearch/compare/3f6b8b84a37a392e639454f20c0e56a661ebe186..34b9d0ea728701b43b5f597484ef8aa2ffa9771d) the unit-tests branch from[ 3f6b8b8 ](https://github.com/opensearch-project/OpenSearch/commit/3f6b8b84a37a392e639454f20c0e56a661ebe186)to[ 34b9d0e ](https://github.com/opensearch-project/OpenSearch/commit/34b9d0ea728701b43b5f597484ef8aa2ffa9771d)1 hour ago

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 1]
```diff
@@ -266,6 +271,121 @@ public void testExtensionsInitialize() throws Exception {
...
+        ExtensionsOrchestrator extensionsOrchestrator = new ExtensionsOrchestrator(settings, extensionDir);
+
+        transportService.registerRequestHandler(
+            ExtensionsOrchestrator.REQUEST_EXTENSION_CLUSTER_STATE,
+            ThreadPool.Names.GENERIC,
+            false,
+            false,
+            ExtensionRequest::new,
+            (request, channel, task) -> {
+                channel.sendResponse(new ClusterStateResponse(clusterName, null, false));
```
Comment (saratvemulapalli) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:352]: Why are we creating a new `ClusterStateResponse` here? 
Ideally we should be able to test the response coming from `extensionsOrchestrator.handleExtensionRequest(request)`

Or Am I missing something here?

Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:303]: Added another test [here](https://github.com/opensearch-project/OpenSearch/pull/3750/files#diff-4e10e710ab4dc067ee8aa4dccb15f340690908cd846d3c8124dc3b17595b9ab4R334) to test the response. 
This test focusses on mocking `registerRequestHandler`

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 2]
```diff
@@ -266,6 +271,121 @@ public void testExtensionsInitialize() throws Exception {
...
+            "firstExtension",
+            "version",
+            "1.0",
+            "opensearch.version",
+            Version.CURRENT.toString(),
+            "java.version",
+            System.getProperty("java.specification.version"),
+            "classname",
+            "FakeExtensions"
+        );
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java]: Was not needed at least for the given test. So removed it.

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 3]
```diff
@@ -266,6 +271,121 @@ public void testExtensionsInitialize() throws Exception {
...
+                "test_node",
+                "test_node",
+                boundAddress.publishAddress(),
+                emptyMap(),
+                emptySet(),
+                Version.CURRENT
+            ),
+            null,
+            Collections.emptySet()
+        );
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java]: Great catch. Created `@Before` and `@After` methods to clean the whole file

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 4]
```diff
@@ -273,13 +294,68 @@ public void testExtensionsInitialize() throws Exception {
...
+                assertNotNull(extensionsOrchestrator.handleExtensionRequest(request));
+            }
+        );
+
+    }
+
+    public void testHandlerResponse() throws Exception {
+        ClusterName clusterName = new ClusterName("cluster-1");
+        ClusterStateResponse clusterStateResponse = new ClusterStateResponse(clusterName, null, false);
+        assertEquals(clusterStateResponse.getClusterName(), clusterName);
```
Comment (saratvemulapalli) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:351]: This is testing the `ClusterStateResponse` which is good, but its not testing the handler response. 

Lets write unit tests for:
- `handleExtensionRequest` pass in a request and verify the response params it sends back.
- Test all responses, eg `ClusterStateResponse` in `ClusterStateResponseTest` file. 

Unit tests are usually for that specific class. Lets write new files for every different class we'd want to test.

Comment (owaiskazi19) [reply server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java]: - Done. Mocked the class type to verify the response is correct
- We already have [ClusterStateResponseTest](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/test/java/org/opensearch/action/admin/cluster/state/ClusterStateResponseTests.java) file present in OpenSearch. Rest of the responses are handled when verifying the class type of return object of class type in handleExtensionRequest. 
- Can you please elaborate more on this?

File: server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java [hunk 5]
```diff
@@ -273,13 +295,34 @@ public void testExtensionsInitialize() throws Exception {
...
+        ExtensionRequest clusterSettingRequest = new ExtensionRequest(
+            ExtensionsOrchestrator.RequestType.REQUEST_EXTENSION_CLUSTER_SETTINGS
+        );
+        assertEquals(extensionsOrchestrator.handleExtensionRequest(clusterSettingRequest).getClass(), ClusterSettingsResponse.class);
+
+        ExtensionRequest localNodeRequest = new ExtensionRequest(ExtensionsOrchestrator.RequestType.REQUEST_EXTENSION_LOCAL_NODE);
+        assertEquals(extensionsOrchestrator.handleExtensionRequest(localNodeRequest).getClass(), LocalNodeResponse.class);
+
+        ExtensionRequest nullRequest = new ExtensionRequest(ExtensionsOrchestrator.RequestType.GET_SETTINGS);
+        assertEquals(extensionsOrchestrator.handleExtensionRequest(nullRequest), null);
```
Comment (saratvemulapalli) [inline server/src/test/java/org/opensearch/extensions/ExtensionsOrchestratorTests.java:322]: Lets not return `null`. Its going to make the system indeterministic when the caller doesnt handle it cleanly. 
Lets throw an exception and make sure this test expects that exception.

**Review by saratvemulapalli — APPROVED** [review]
Thanks @owaiskazi19 for these changes!!

File: server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java [hunk 6]
```diff
@@ -107,10 +107,14 @@ public ExtensionsOrchestrator(Settings settings, Path extensionsPath) throws IOE
 
     public void setTransportService(TransportService transportService) {
         this.transportService = transportService;
+        registerRequestHandler();
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsOrchestrator.java:110]: Sure. Will cover it in the next PR :)

**Pull request merged**